### PR TITLE
Determine version information from version.json for newer Minecraft versions

### PR DESCRIPTION
**Pull request by andreblanke**
Changes the logic for the detection of the Minecraft and server protocol versions to include a check for the server JAR's [version.json](https://minecraft.fandom.com/wiki/Version.json) that is present in newer Minecraft versions. With these changes the new logic will be as follows:

1. If config values for protocol and version are present, use those
2. Otherwise, try to fetch the version information from the `version.json`
    - Log an error to inform the user that the versions have to be manually set if the file cannot be found or parsing it failed
3. Only update the config version information values in `getServInfo` for older versions of Minecraft before `version.json` was introduced (namely versions before snapshot 18w47b, checked by comparing the protocol version against 447)

The changes have mostly been tested except for the one in `getServInfo` but I hope to get around that on the weekend. Sorry that it's been a while since our conversation.

### CODE REVIEW

**Comment by gekigek99** [comment]
Thanks for the pr! This change is probably eligible for a msh version update.

No worries for the late response, I'll wait for the test results of the weekend! ;)

**Comment by andreblanke** [comment]
Should be good to go. Turns out the changes made to `getServInfo` were not necessary after all. For some reason I assumed the config file would be updated with the information

File: lib/config/config.go [hunk 1]
```diff
@@ -84,6 +88,23 @@ func LoadConfig() *errco.Error {
...
+	// set server and protocol version from server JAR file if not manually set in the config.
+	// required for backward-compatibility and for Minecraft versions without a version.info inside the JAR file
+	// (see https://minecraft.fandom.com/wiki/Version.json)
+	if ConfigRuntime.Server.Version == "" || ConfigRuntime.Server.Protocol == 0 {
+		var version, protocol, errMsh = getVersionInfo()
+		if errMsh != nil {
+			return errMsh.AddTrace("LoadConfig")
+		}
+
+		if ConfigRuntime.Server.Version == "" {
```
Comment (gekigek99) [inline lib/config/config.go:100]: why are you checking them again?
couldn't we just set them without an if?

File: lib/config/config.go [hunk 2]
```diff
@@ -84,6 +88,23 @@ func LoadConfig() *errco.Error {
 
 	errco.Logln(errco.LVL_D, "msh proxy setup: %s:%d --> %s:%d", ListenHost, ListenPort, TargetHost, TargetPort)
 
+	// set server and protocol version from server JAR file if not manually set in the config.
+	// required for backward-compatibility and for Minecraft versions without a version.info inside the JAR file
```
Comment (gekigek99) [inline lib/config/config.go:92]: what is version.info used for?
what are the differences with Version.json?

File: lib/config/config.go [hunk 3]
```diff
@@ -214,3 +226,54 @@ func (c *Configuration) loadRuntime(base *Configuration) *errco.Error {
...
+			return errco.NewErr(errco.ERROR_VERSION_LOAD, errco.LVL_D, "extractVersionInfo", err.Error())
+		}
+
+		var info model.VersionInfo
+		err = json.Unmarshal(versionsBytes, &info)
+		if err != nil {
+			return errco.NewErr(errco.ERROR_VERSION_LOAD, errco.LVL_D, "extractVersionInfo", err.Error())
+		}
+
+		if c.Server.Version == "" {
```
Comment (gekigek99) [inline lib/config/config.go:267]: why are you checking them again?
couldn't we just set them without an if?

File: lib/config/config.go [hunk 4]
```diff
@@ -84,6 +88,14 @@ func LoadConfig() *errco.Error {
 
 	errco.Logln(errco.LVL_D, "msh proxy setup: %s:%d --> %s:%d", ListenHost, ListenPort, TargetHost, TargetPort)
 
+	// set server and protocol version from server JAR file if not specified in config/arguments.
+	// required for backward-compatibility and for minecraft versions without a version.info inside the JAR file.
```
Comment (gekigek99) [inline lib/config/config.go:92]: what is version.info used for?
what are the differences with Version.json?

**Comment by gekigek99** [comment]
do you think it still works?

File: lib/config/config.go (hunk 4 above)
Comment (andreblanke) [reply lib/config/config.go]: Whoops, sorry. That should've been `version.json` as well

**Comment by andreblanke** [comment]
It's still working fine for me but I wouldn't write the new version information to the config file and instead read it from the JAR file each time. Otherwise the configuration settings might get out of sync, for example when the server is updated

File: lib/config/config.go (hunk 4 above)
Comment (gekigek99) [reply lib/config/config.go]: why is it required for `backward-compatibility and for minecraft versions without a version.json inside the JAR file` ?

**Comment by gekigek99** [comment]
why is it required for `backward-compatibility and for minecraft versions without a version.json inside the JAR file` ?

**Comment by andreblanke** [comment]
The if statement itself was meant by that. It basically allows you as a user to keep using the old logic of MSH and to skip the check of `version.json` if both version and protocol are defined (via the config or the arguments), basically overriding the values of the server JAR.

A (maybe better, now that I think about it) alternative would be to always check for the presence of `version.json` and, if the version information can be successfully determined, removing the version information from the config file. If the version file cannot be found the version information is kept inside the config file and used as usual

**Pull request merged**